articles/tests/test_views.py

```diff
@@ -260,3 +263,186 @@ def test_get_article_with_invalid_slug_fail(self) -> None:
+        self.assertEqual(len(response.data.get("likes")), likes + 1)
+        self.assertEqual(len(response.data.get("dislikes")), dislikes)
+
+    def test_dislike_article_pass(self) -> None:
+        """test the ability to dislike an article"""
+        dislikes = self.article.dislikes.count()
+        response = self.client.patch(
+            reverse("article-unfavorite", kwargs={"slug": self.article.slug}),
+            **self.bearer_token,
+        )
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        self.assertEqual(len(response.data.get("dislikes")), dislikes + 1)
+        self.assertTrue(response.data.get("unfavorited"))
+        self.assertFalse(response.data.get("favorited"))
+
+    def test_undislike_article_pass(self) -> None:
+        """test the ability to undislike an article. Removes user from the dislikes list"""
+        dislikes = self.article.dislikes.count()
+        response = self.client.patch(
+            reverse("article-unfavorite", kwargs={"slug": self.article.slug}),
+            **self.bearer_token,
+        )
+        response = self.client.patch(
+            reverse("article-unfavorite", kwargs={"slug": self.article.slug}),
+            **self.bearer_token,
+        )
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        self.assertEqual(len(response.data.get("dislikes")), dislikes)
+        self.assertFalse(response.data.get("unfavorited"))
+
+    def test_like_article_without_authentication_fail(self) -> None:
+        """test inability to like an article without authentication"""
+        response = self.client.patch(
+            reverse("article-favorite", kwargs={"slug": self.article.slug}),
+        )
+        self.assertEqual(response.status_code, status.HTTP_401_UNAUTHORIZED)
+        self.assertEqual(
+            json.loads(response.content).get("detail"),
+            "Authentication credentials were not provided.",
+        )
+        self.assertRaisesMessage(
+            PermissionDenied, "Authentication credentials were not provided."
+        )
+
+    def test_dislike_favorited_article_removes_the_like_entry_pass(
+        self,
+    ) -> None:
+        """
+        test that a dislike on a favorited article removes the like entry
+        If a user has already liked an article, and then dislikes it, the like entry should be removed
+        before being added to the dislike entry
+        """
+        likes = self.article.likes.count()
+        dislikes = self.article.dislikes.count()
+        response = self.client.patch(
+            reverse("article-favorite", kwargs={"slug": self.article.slug}),
+            **self.bearer_token,
+        )
+        response = self.client.patch(
+            reverse("article-unfavorite", kwargs={"slug": self.article.slug}),
+            **self.bearer_token,
+        )
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        self.assertEqual(len(response.data.get("likes")), likes)
```

Comment:
Shouldn't the likes decrease because it has removed that entry?

Reply:
The first response registers a like which is the removed automatically when the same user dislikes the same article. So the likes increase by one then reduce by one this remaining constant